### PR TITLE
[many dialects] Add support for SQLAlchemy templating

### PR DESCRIPTION
**Pull request by jacopofar**
### Brief summary of the change made
This closes in #1216 by adding support for the SQLAlchemy parameter passing. SQLALchemy is a very common library to interface with DBs and it supports multiple engines (e.g. Postgres, MySQL, SQLite, etc.)

### Are there any other side effects of this change that we should be aware of?
Not that I can see, the usage is identical to the other templaters but much simpler

### Pull Request checklist
- [X ] Please confirm you have completed any of the necessary steps below.

### CODE REVIEW

**Comment by WittierDinosaur** [comment]
Probably one for @barrywhart or @alanmcruickshank to review

File: src/sqlfluff/core/templaters/sqlalchemy.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+            raw_slices.append(
+                RawFileSlice(
+                    raw=in_str[last_pos_raw : span[0]],
+                    slice_type="literal",
+                    source_idx=last_pos_raw,
+                )
+            )
+            out_str += in_str[last_pos_raw : span[0]]
+            # add the current replaced element
+            start_teplate_pos = last_pos_templated + last_literal_length
```
Comment (barrywhart) [inline src/sqlfluff/core/templaters/sqlalchemy.py:118]:
```suggestion
            start_template_pos = last_pos_templated + last_literal_length
```

File: src/sqlfluff/core/templaters/sqlalchemy.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+            template_slices.append(
+                TemplatedFileSlice(
+                    slice_type="block_start",
+                    source_slice=slice(span[0], span[1], None),
+                    templated_slice=slice(
+                        start_teplate_pos, start_teplate_pos + len(replacement), None
```
Comment (barrywhart) [inline src/sqlfluff/core/templaters/sqlalchemy.py:124]:
```suggestion
                        start_template_pos, start_template_pos + len(replacement), None
```

File: src/sqlfluff/core/templaters/sqlalchemy.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+                RawFileSlice(
+                    raw=in_str[span[0] : span[1]],
+                    slice_type="block_start",
+                    source_idx=span[0],
+                )
+            )
+            out_str += replacement
+            # update the indexes
+            last_pos_raw = span[1]
+            last_pos_templated = start_teplate_pos + len(replacement)
```
Comment (barrywhart) [inline src/sqlfluff/core/templaters/sqlalchemy.py:138]:
```suggestion
            last_pos_templated = start_template_pos + len(replacement)
```

**Review by barrywhart — APPROVED** [review]
Other than a typo in a variable name, looks great!

Hypothetically, how much effort would it be to make this templater more generic, e.g. so we could subclass it to handle other templating syntax? For example, see #1871.

**Review by barrywhart — CHANGES_REQUESTED** [review]
Changing review to "request changes" so we can fix the typos before merging.

File: src/sqlfluff/core/templaters/sqlalchemy.py (hunk 1 above)
Comment (jacopofar) [reply src/sqlfluff/core/templaters/sqlalchemy.py]: fixed

File: src/sqlfluff/core/templaters/sqlalchemy.py (hunk 3 above)
Comment (jacopofar) [reply src/sqlfluff/core/templaters/sqlalchemy.py]: done

**Review by barrywhart — APPROVED** [review]

**Comment by barrywhart** [comment]
@jacopofar: If you're up for making that change, that would be great! Do you think it'd be better to do that in this or in a separate PR?

**Comment by jacopofar** [comment]
we can close this and I'll open a separate request for that, should be easy to change

**Comment by barrywhart** [comment]
Ok, merging now. For reference, issue #1871 would hopefully be easy to implement after making this templater more generic.

**Comment by jacopofar** [comment]
oh sorry by close I meant without merging, since the new one will replace it completely and break the (unlikely) code already using it

**Comment by barrywhart** [comment]
Oh, I'm sorry. Well, it's not in a release yet, so it's fine to change. It should be possible (assuming we want to) to keep it compatible (I.e. same templater name in the config file).

**Pull request merged**